app/modules/ads/module.js

```diff
@@ -0,0 +1,553 @@
+		} else if (this.uapResult) {
+			uapCallback();
+		} else {
+			noUapCallback();
+		}
+	}
+
+	/**
+	 * Method for sampling and pushing ads-related events
+	 * @arguments coming from ads tracking request
+	 * It's called by track() method in wikia.tracker fetched from app by ads code
+	 *
+	 * @returns {void}
+	 */
+	static gaTrackAdEvent() {
+		// Percentage of all the track requests to go through
+		const adHitSample = 1;
+
+		// Sampling on GA side will kill the performance as we need to allocate object each time we track
+		// ToDo: Optimize object allocation for tracking all events
+		if (Math.random() * 100 <= adHitSample) {
+			M.tracker.UniversalAnalytics.trackAds(...arguments);
+		}
+	}
+
+	/**
+	 * Track pageview in Krux (imported from Oasis/MediaWiki)
+	 *
+	 * mobileId variable is the ID referencing to the mobile site
+	 * (see ads_run.js and krux.js in app repository)
+	 *
+	 * @returns {void}
+	 */
+	trackKruxPageView() {
+		// if (this.krux && typeof this.krux.load === 'function') {
+		// 	console.info('Track pageView: Krux');
+
+		// 	// @todo XW-123 add logging to kibana how many times failed to load
+		// 	this.krux.load(config.tracking.krux.mobileId);
+		// }
+	}
+
+	/**
+	 * @param {String} name
+	 * @param {Object} GAOption
+	 * @param {Boolean} isAdBlockDetected
+	 * @returns {void}
+	 */
+	trackBlocking(name, GAOption, isAdBlockDetected) {
+		let value = isAdBlockDetected ? 'Yes' : 'No';
+
+		Ads.setPreviousDetectionResult(name, isAdBlockDetected);
+		M.tracker.UniversalAnalytics.setDimension(GAOption.dimension, value);
+		M.tracker.UniversalAnalytics.track(`ads-${GAOption.name}-detection`, 'impression', value, 0, true);
+
+		Ads.gaTrackAdEvent.call(this, `ad/${GAOption.name}/detection`, value, '', 0, true);
+	}
+
+	static setPreviousDetectionResult(name, isAdBlockDetected) {
+		Ads.previousDetectionResults[name].value = isAdBlockDetected;
+		Ads.previousDetectionResults[name].exists = true;
+	}
+
+	/**
+	 * @returns {void}
+	 */
+	addDetectionListeners() {
+		const GASettings = this.GASettings,
+			listenerSettings = [
+				{
+					name: 'babDetector',
+					eventName: 'bab.blocking',
+					value: true,
+				},
+				{
+					name: 'babDetector',
+					eventName: 'bab.not_blocking',
+					value: false,
+				},
+			];
+
+		listenerSettings.map((listenerSetting) => {
+			document.addEventListener(listenerSetting.eventName, () => {
+				this.trackBlocking(listenerSetting.name, GASettings[listenerSetting.name], listenerSetting.value);
+			});
+		});
+	}
+
+	/**
+	 * @param {*} adsContext
+	 * @returns {void}
+	 */
+	setContext(adsContext) {
+		this.adsContext = adsContext ? adsContext : null;
+	}
+
+	getTargetingValue(key) {
+		return this.adsContext && this.adsContext.targeting && this.adsContext.targeting[key];
+	}
+
+	/**
+	 * Turns off all ads for logged in user
+	 *
+	 * @param {*} adsContext
+	 * @returns {void}
+	 */
+	turnOffAdsForLoggedInUsers(adsContext) {
+		// TODO: Refactor/remove while working on ADEN-2189
```

Comment:
ADEN-2189 is a won't fix. let's just face it - we won't do it until it's needed. let's remove the TODO.

Reply:
ok, chef.